backend/canisters/group/impl/src/updates/c2c_invite_users.rs

```diff
@@ -0,0 +1,98 @@
+use crate::activity_notifications::handle_activity_notification;
+use crate::guards::caller_is_user_index_or_local_user_index;
+use crate::model::invited_users::UserInvitation;
+use crate::{mutate_state, run_regular_jobs, RuntimeState};
+use canister_api_macros::update_msgpack;
+use canister_tracing_macros::trace;
+use chat_events::ChatEventInternal;
+use group_canister::c2c_invite_users::{Response::*, *};
+use types::{EventIndex, MessageIndex, UsersInvited};
+
+const MAX_INVITES: usize = 100;
+
+#[update_msgpack(guard = "caller_is_user_index_or_local_user_index")]
+#[trace]
+fn c2c_invite_users(args: Args) -> Response {
+    run_regular_jobs();
+
+    mutate_state(|state| c2c_invite_users_impl(args, state))
+}
+
+fn c2c_invite_users_impl(args: Args, runtime_state: &mut RuntimeState) -> Response {
+    if runtime_state.data.is_frozen() {
+        return ChatFrozen;
+    }
+
+    let now = runtime_state.env.now();
+
+    if let Some(participant) = runtime_state.data.participants.get_by_user_id(&args.caller) {
+        // The original caller must be authorized to invite other users
+        if participant.suspended.value
+            || (!runtime_state.data.is_public && !participant.role.can_invite_users(&runtime_state.data.permissions))
+        {
+            return NotAuthorized;
+        }
+
+        // Filter out users who are already members
+        let invited_users: Vec<_> = args
+            .users
+            .iter()
+            .filter(|(_, principal)| runtime_state.data.participants.get_by_principal(principal).is_none())
+            .copied()
+            .collect();
+
+        let user_ids: Vec<_> = invited_users.iter().map(|(user_id, _)| user_id).copied().collect();
+
+        if !runtime_state.data.is_public {
```

Comment:
Nothing actually happens if the group is public... so surely we can short circuit earlier or reject the call?